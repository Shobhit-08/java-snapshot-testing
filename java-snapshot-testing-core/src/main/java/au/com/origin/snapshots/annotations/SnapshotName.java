package au.com.origin.snapshots.annotations;

import java.lang.annotation.*;

@Documented
@Target({ElementType.METHOD})
@Inherited
@Retention(RetentionPolicy.RUNTIME)
public @interface SnapshotName {
  String value();
}
