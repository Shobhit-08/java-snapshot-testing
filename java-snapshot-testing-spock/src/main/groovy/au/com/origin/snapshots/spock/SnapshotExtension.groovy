package au.com.origin.snapshots.spock

import au.com.origin.snapshots.SnapshotVerifier
import au.com.origin.snapshots.config.PropertyResolvingSnapshotConfig
import au.com.origin.snapshots.config.SnapshotConfig
import au.com.origin.snapshots.config.SnapshotConfigInjector
import org.spockframework.runtime.extension.AbstractAnnotationDrivenExtension
import org.spockframework.runtime.model.SpecInfo

class SnapshotExtension extends AbstractAnnotationDrivenExtension<EnableSnapshots> implements SnapshotConfigInjector {

    SnapshotVerifier snapshotVerifier;

    void visitSpecAnnotation(EnableSnapshots annotation, SpecInfo spec) {
        this.snapshotVerifier = new SnapshotVerifier(getSnapshotConfig(), spec.reflection, false)
    }

    void visitSpec(SpecInfo spec) {
        def snapshotMethodInterceptor = new SnapshotMethodInterceptor(snapshotVerifier)
        spec.allFeatures.featureMethod*.addInterceptor(snapshotMethodInterceptor)
        spec.addCleanupSpecInterceptor(snapshotMethodInterceptor)
    }

    @Override
    SnapshotConfig getSnapshotConfig() {
        return new PropertyResolvingSnapshotConfig()
    }
}