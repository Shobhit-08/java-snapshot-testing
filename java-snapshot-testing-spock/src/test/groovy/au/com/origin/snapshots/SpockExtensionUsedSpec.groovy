package au.com.origin.snapshots

import au.com.origin.snapshots.annotations.SnapshotName
import au.com.origin.snapshots.spock.EnableSnapshots
import spock.lang.Specification
import spock.lang.Unroll

@EnableSnapshots
class SpockExtensionUsedSpec extends Specification {

    Expect expect

    @SnapshotName("Should use extension")
    def "Should use extension"(Expect expect) {
        when:
        expect.toMatchSnapshot("Hello World")

        then:
        true
    }

    @SnapshotName("Should use extension again")
    def "Should use extension again"(Expect expect) {
        when:
        expect.toMatchSnapshot("Hello World")

        then:
        true
    }

    @SnapshotName("Should use extension via instance variable")
    def "Should use extension via instance variable"() {
        when:
        expect.toMatchSnapshot("Hello World")

        then:
        true
    }

    @SnapshotName("DataTable example 1")
    @Unroll
    def 'DataTable example 1: #letter'(def letter) {
        given: 'I use an @Unroll function'
        String result = letter.toUpperCase()

        when: 'I snapshot the letter'
        expect.scenario("letter $letter").toMatchSnapshot(result)

        then:
        true

        where:
        [letter] << [['A'],['B'],['C']]
    }


    @SnapshotName("DataTable example 2")
    def 'DataTable example 2: #scenario to uppercase'() {
        when: 'I convert to uppercase'
        String result = value.toUpperCase();
        then: 'Should convert letters to uppercase'
        // Check you snapshot against your output using a unique scenario
        expect.scenario(scenario).toMatchSnapshot(result)
        where:
        scenario | value
        'letter' | 'a'
        'number' | '1'
    }

    @SnapshotName("Can run a non snapshot test")
    def "Can run a non snapshot test"() {
        when:
        def isTrue = true

        then:
        isTrue
    }

}
