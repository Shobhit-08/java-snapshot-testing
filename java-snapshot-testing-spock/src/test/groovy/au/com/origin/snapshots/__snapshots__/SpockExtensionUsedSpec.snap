DataTable example 1[letter A]=[
A
]


DataTable example 1[letter B]=[
B
]


DataTable example 1[letter C]=[
C
]


DataTable example 2[letter]=[
A
]


DataTable example 2[number]=[
1
]


Should use extension again=[
Hello World
]


Should use extension via instance variable=[
Hello World
]


Should use extension=[
Hello World
]